diff -ruN a/file_index.cc b/file_index.cc
--- a/file_index.cc	2025-10-05 01:03:50.266789887 +0000
+++ b/file_index.cc	2025-10-05 01:03:08.914166910 +0000
@@ -25,6 +25,7 @@
 #include <vector>
 #include <stdint.h>
 #include <unistd.h>
+#include <snprintf.h>
 
 #include "lzip.h"
 #include "file_index.h"
diff -ruN a/main.cc b/main.cc
--- a/main.cc	2025-10-05 01:03:49.826783258 +0000
+++ b/main.cc	2025-10-05 01:03:10.402189327 +0000
@@ -53,6 +53,7 @@
 #if defined(__OS2__)
 #include <io.h>
 #endif
+#include <snprintf.h>
 
 #include "arg_parser.h"
 #include "lzip.h"
