diff -ruN a/serverloop.c b/serverloop.c
--- a/serverloop.c	2004-05-17 00:29:05.000000000 +0000
+++ b/serverloop.c	2025-09-21 21:32:24.763626706 +0000
@@ -139,7 +139,12 @@
   int i, i1;
   int accept_fd;
   struct sockaddr addr;
-  SOCKLEN_T addrlen;
+  #if defined(__sun) && (__SUNOS_RELEASE_ > 551) // Solaris 2.6 and newer
+      socklen_t addrlen;
+  #else // Solaris 2.5.1 and older
+      int addrlen;
+  #endif
+
   int num, numbytes, error;
   int pipedes[2];
   time_t last_seed_stat = 0, last_seed_ext = 0;
@@ -316,7 +321,7 @@
       }
       if (FD_ISSET(service_socket[i], &rfds))
       {
-        addrlen = (SOCKLEN_T) sizeof(addr);
+        addrlen = sizeof(addr);
         accept_fd = accept(service_socket[i], &addr, &addrlen);
         if (accept_fd < 0)
 	  continue;		/*
diff -ruN a/tcp_socket.c b/tcp_socket.c
--- a/tcp_socket.c	2002-03-10 08:32:58.000000000 +0000
+++ b/tcp_socket.c	2025-09-21 21:25:55.849364253 +0000
@@ -105,13 +105,13 @@
   non_blocking(sockfd, non_block);
   close_on_exec(sockfd, close_exec);
 
-  if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &on,
-		 (SOCKLEN_T) sizeof(on)) < 0) {
+  if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, (const void *) &on,
+		 sizeof(on)) < 0) {
     error_fatal("Could not set SO_REUSEADDR", errno);
     exit(EXIT_FAILURE);
   }
-  if (setsockopt(sockfd, SOL_SOCKET, SO_KEEPALIVE, &on,
-		 (SOCKLEN_T) sizeof(on)) < 0) {
+  if (setsockopt(sockfd, SOL_SOCKET, SO_KEEPALIVE, (const void *) &on,
+		 sizeof(on)) < 0) {
     error_fatal("Could not set SO_KEEPALIVE", errno);
     exit(EXIT_FAILURE);
   }
